cylc/flow/xtriggers/workflow_state.py

```diff
@@ -14,18 +14,117 @@
 # along with this program.  If not, see <http://www.gnu.org/licenses/>.
 
-import sqlite3
-from typing import Dict, Optional, Tuple
+from typing import Dict, Optional, Tuple, Any
+import asyncio
+from inspect import signature
 
-from metomi.isodatetime.parsers import TimePointParser
-
-from cylc.flow.cycling.util import add_offset
-from cylc.flow.dbstatecheck import CylcWorkflowDBChecker
-from cylc.flow.pathutil import get_cylc_run_dir
-from cylc.flow.workflow_files import infer_latest_run_from_id
+from cylc.flow.scripts.workflow_state import WorkflowPoller
+from cylc.flow.id import tokenise
+from cylc.flow.exceptions import WorkflowConfigError
+from cylc.flow.task_state import TASK_STATUS_SUCCEEDED
 
 
 def workflow_state(
+    workflow_task_id: str,
+    offset: Optional[str] = None,
+    flow_num: Optional[int] = 1,
+    is_output: bool = False,
+    is_message: bool = False,
+    alt_cylc_run_dir: Optional[str] = None,
+) -> Tuple[bool, Dict[str, Any]]:
+    """Connect to a workflow DB and check a task status or output.
+
+    If the status or output has been achieved, return {True, result}.
+
+    Arg:
+        workflow_task_id:
+            ID (workflow//point/task:selector) of the target task.
+        offset:
+            Offset from cycle point as an ISO8601 or integer duration,
+            e.g. PT1H (1 hour) or P1 (1 integer cycle)
+        flow_num:
+            Flow number of the target task.
+        alt_cylc_run_dir:
+            Alternate cylc-run directory, e.g. for another user.
+
+    Returns:
+        tuple: (satisfied, result)
+
+        satisfied:
+            True if ``satisfied`` else ``False``.
+        result:
+            Dict {workflow_id, task_id, task_selector, flow_num}.
+
+    """
+    poller = WorkflowPoller(
+        workflow_task_id,
+        offset,
+        flow_num,
+        alt_cylc_run_dir,
+        TASK_STATUS_SUCCEEDED,
+        is_output, is_message,
+        old_format=False,
+        condition=workflow_task_id,
+        max_polls=1,  # (for xtriggers the scheduler does the polling)
+        interval=0,  # irrelevant for 1 poll
+        args=[]
+    )
+    if asyncio.run(poller.poll()):
+        return (
+            True,
+            {
+                "workflow_id": poller.workflow_id,
+                "task_id": f"{poller.cycle}/{poller.task}",
+                "task_selector": poller.selector,
+                "flow_num": poller.flow_num
+            }
+        )
+    else:
+        return (
+            False,
+            {}
+        )
+
+
+def validate(args: Dict[str, Any]):
+    """Validate workflow_state xtrigger function args.
+
+    Arguments:
+        workflow_task_id:
+            full workflow//cycle/task[:selector]
+        offset:
+            must be a valid status
+        flow_num:
+            must be an integer
+        alt_cylc_run_dir:
+            must be a valid path
+
+    """
+    try:
+        tokens = tokenise(args["workflow_task_id"])
+    except KeyError:
```

Comment:
What input was getting a `KeyError` from this?

Reply:
Fair question. I think that pre-dates the new function signature introspection. Removing it...